### PR TITLE
move walkRoute to dispatch package.

### PR DESCRIPTION
**Pull request by johncming**
move walkRoute to dispatch package for better encapsulation and add test case.

### CODE REVIEW

File: dispatch/route_test.go [hunk 1]
```diff
@@ -267,3 +267,90 @@ routes:
...
+
+	var ctree config.Route
+	if err := yaml.UnmarshalStrict([]byte(in), &ctree); err != nil {
+		t.Fatal(err)
+	}
+	tree := NewRoute(&ctree, nil)
+
+	var got, expected []string
+
+	// DFS
```
Comment (simonpasquier) [inline dispatch/route_test.go]: This can be removed

File: dispatch/route_test.go [hunk 2]
```diff
@@ -267,3 +267,90 @@ routes:
...
+      group_wait: 2m
+`
+
+	var ctree config.Route
+	if err := yaml.UnmarshalStrict([]byte(in), &ctree); err != nil {
+		t.Fatal(err)
+	}
+	tree := NewRoute(&ctree, nil)
+
+	var got, expected []string
```
Comment (simonpasquier) [inline dispatch/route_test.go]: lets move `var got |]string` right before calling `tree.Walk()`.

File: dispatch/route_test.go [hunk 3]
```diff
@@ -267,3 +267,90 @@ routes:
...
+
+	// DFS
+	expected = []string{
```
Comment (simonpasquier) [inline dispatch/route_test.go]: `expected = []string{` -> `expected := []string{`

File: dispatch/route_test.go (hunk 1 above)
Comment (johncming) [reply dispatch/route_test.go]: ok.

File: dispatch/route_test.go (hunk 2 above)
Comment (johncming) [reply dispatch/route_test.go]: ok.

File: dispatch/route_test.go (hunk 3 above)
Comment (johncming) [reply dispatch/route_test.go]: ok

**Comment by simonpasquier** [comment]
The DCO is failing.

File: dispatch/route_test.go (hunk 1 above)
Comment (simonpasquier) [reply dispatch/route_test.go]: still to be removed :smirk:

**Comment by johncming** [comment]
> The DCO is failing.

@simonpasquier fixed

**Comment by simonpasquier** [comment]
Please remove this [comment](https://github.com/prometheus/alertmanager/pull/2136/files#diff-8d3bcd8a072f2a0eb4586f92a18fa168R334) and it will be good to go. Thanks!

**Comment by simonpasquier** [comment]
Thanks!

**Pull request merged**